File: plugins/plugin-github/typescript/actions/createPullRequest.ts
import {
  type Action,
  type ActionExample,
  type ActionResult,
  type Content,
  type HandlerCallback,
  type HandlerOptions,
  type IAgentRuntime,
  logger,
  type Memory,
  type State,
} from "@elizaos/core";
import { requireActionSpec } from "../generated/specs/spec-helpers";
import { GITHUB_SERVICE_NAME, type GitHubService } from "../service";
import { type CreatePullRequestParams, createPullRequestSchema, formatZodErrors } from "../types";

const spec = requireActionSpec("CREATE_PULL_REQUEST");

const examples: ActionExample[][] = [
  [
    {
      name: spec.name,
      content: {
        text: "Create a pull request from feature/dark-mode to main with title 'Add dark mode support'",
      },
    },
    {
      name: "{{agent}}",
      content: {
        text: "I'll create a pull request from feature/dark-mode to main.",
        actions: ["CREATE_GITHUB_PULL_REQUEST"],
      },
    },
  ],
  [
    {
      name: "{{user1}}",
      content: {
        text: "Open a PR to merge my-branch into develop",
      },
    },
    {
      name: "{{agent}}",
      content: {
        text: "Creating a pull request to merge my-branch into develop.",
        actions: ["CREATE_GITHUB_PULL_REQUEST"],
      },
    },
  ],
];

export const createPullRequestAction: Action = {
  name: "CREATE_GITHUB_PULL_REQUEST",
  similes: spec.similes ? [...spec.similes] : [],
  description: spec.description,

  validate: async (runtime: IAgentRuntime, message: Memory, _state?: State): Promise<boolean> => {
    const service = runtime.getService(GITHUB_SERVICE_NAME);
    if (!service) {
      return false;
    }

    const text = (message.content as Content).text?.toLowerCase() ?? "";
    return text.includes("pull request") || text.includes("pr") || text.includes("merge");
  },

  handler: async (
    runtime: IAgentRuntime,
    message: Memory,
    state: State | undefined,
    _options?: HandlerOptions,
    callback?: HandlerCallback
  ): Promise<ActionResult> => {
    const service = runtime.getService<GitHubService>(GITHUB_SERVICE_NAME);

    if (!service) {
      logger.error("GitHub service not available");
      if (callback) {
        await callback({
          text: "GitHub service is not available. Please ensure the plugin is properly configured.",
        });
      }
      return { success: false };
    }

    try {
      const content = message.content as Content;
      const text = content.text ?? "";

      const params: CreatePullRequestParams = {
        owner: (state?.owner as string) ?? service.getConfig().owner ?? "",
        repo: (state?.repo as string) ?? service.getConfig().repo ?? "",
        title: (state?.title as string) ?? text.slice(0, 100),
        body: (state?.body as string) ?? text,
        head: (state?.head as string) ?? "",
        base: (state?.base as string) ?? service.getConfig().branch ?? "main",
        draft: (state?.draft as boolean) ?? false,
      };

      const validation = createPullRequestSchema.safeParse(params);
      if (!validation.success) {
        const errors = formatZodErrors(validation.error);
        logger.error(`Invalid pull request parameters: ${errors}`);
        if (callback) {
          await callback({
            text: `I couldn't create the pull request due to missing information: ${errors}`,
          });
        }
        return { success: false };
      }

      const pr = await service.createPullRequest(params);

      logger.info(`Created pull request #${pr.number}: ${pr.title}`);

      if (callback) {
        await callback({
          text: `Created pull request #${pr.number}: "${pr.title}"\n\nFrom: ${pr.head.ref}\nTo: ${pr.base.ref}\n\nView it at: ${pr.htmlUrl}`,
        });
      }

      return { success: true };
    } catch (error) {
      const errorMessage = error instanceof Error ? error.message : "Unknown error";
      logger.error(`Failed to create pull request: ${errorMessage}`);

      if (callback) {
        await callback({
          text: `Failed to create the pull request: ${errorMessage}`,
        });
      }

      return { success: false };
    }
  },

  examples,
};

export default createPullRequestAction;

File: plugins/plugin-github/rust/src/actions/create_pull_request.rs
#![allow(missing_docs)]

use async_trait::async_trait;
use serde_json::json;

use super::{ActionContext, ActionResult, GitHubAction};
use crate::error::Result;
use crate::types::CreatePullRequestParams;
use crate::GitHubService;

pub struct CreatePullRequestAction;

#[async_trait]
impl GitHubAction for CreatePullRequestAction {
    fn name(&self) -> &str {
        "CREATE_GITHUB_PULL_REQUEST"
    }

    fn description(&self) -> &str {
        "Creates a new pull request in a GitHub repository to merge changes from one branch to another."
    }

    fn similes(&self) -> Vec<&str> {
        vec![
            "OPEN_PR",
            "CREATE_PR",
            "NEW_PULL_REQUEST",
            "SUBMIT_PR",
            "OPEN_PULL_REQUEST",
            "MERGE_REQUEST",
        ]
    }

    async fn validate(&self, context: &ActionContext) -> Result<bool> {
        let text = context
            .message
            .get("content")
            .and_then(|c| c.get("text"))
            .and_then(|t| t.as_str())
            .unwrap_or("")
            .to_lowercase();

        Ok(text.contains("pull request") || text.contains("pr") || text.contains("merge"))
    }

    async fn handler(
        &self,
        context: &ActionContext,
        service: &GitHubService,
    ) -> Result<ActionResult> {
        let text = context
            .message
            .get("content")
            .and_then(|c| c.get("text"))
            .and_then(|t| t.as_str())
            .unwrap_or("");

        let head = context
            .state
            .get("head")
            .and_then(|h| h.as_str())
            .unwrap_or("feature");

        let base = context
            .state
            .get("base")
            .and_then(|b| b.as_str())
            .unwrap_or(&service.config().branch);

        let title = context
            .state
            .get("title")
            .and_then(|t| t.as_str())
            .map(|s| s.to_string())
            .unwrap_or_else(|| text.chars().take(100).collect());

        let params = CreatePullRequestParams {
            owner: context.owner.clone(),
            repo: context.repo.clone(),
            title,
            body: Some(text.to_string()),
            head: head.to_string(),
            base: base.to_string(),
            draft: false,
            maintainer_can_modify: true,
        };

        let pr = service.create_pull_request(params).await?;

        Ok(ActionResult::success(
            format!("Created pull request #{}: {}", pr.number, pr.title),
            json!({
                "pull_number": pr.number,
                "html_url": pr.html_url,
                "head": pr.head.branch_ref,
                "base": pr.base.branch_ref,
            }),
        ))
    }
}
